packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/queue/queue.ts

```diff
@@ -0,0 +1,384 @@
+            throw Error(`Cannot configure auto scaling only for the subscription queue of service '${this.parentService.id}'. Need to specify 'eventsQueueDelay'.`);
+          }
+          this.subscriptionQueues.add(subsQueue);
+        }
+      }
+    }
+  }
+
+  /**
+   * Add hooks to the main application extension so that it is modified to
+   * add the events queue URL to the container environment.
+   */
+  public addHooks() {
+    const container = this.parentService.serviceDescription.get('service-container') as Container;
+
+    if (!container) {
+      throw new Error('Queue Extension requires an application extension');
+    }
+
+    container.addContainerMutatingHook(new QueueExtensionMutatingHook({
+      environment: this.environment,
+    }));
+  }
+
+  /**
+   * After the task definition has been created, this hook grants SQS permissions to the task role.
+   *
+   * @param taskDefinition The created task definition
+   */
+  public useTaskDefinition(taskDefinition: ecs.TaskDefinition) {
+    this._eventsQueue.grantConsumeMessages(taskDefinition.taskRole);
+    for (const subsQueue of this.subscriptionQueues) {
+      subsQueue.queue.grantConsumeMessages(taskDefinition.taskRole);
+    }
+  }
+
+  /**
+   * When this hook is implemented by extension, it allows the extension
+   * to use the service which has been created. It is used to add target tracking
+   * scaling policies for the SQS Queues of the service. It also creates an AWS Lambda
+   * Function for calculating the backlog per task metric.
+   *
+   * @param service - The generated service.
+   */
+  public useService(service: ecs.Ec2Service | ecs.FargateService) {
+    if (!this._eventsQueueDelay) {
+      return;
+    }
+    if (!this.parentService.scalableTaskCount) {
+      throw Error(`Auto scaling target for the service '${this.parentService.id}' hasn't been configured. Please use Service construct to configure 'minTaskCount' and 'maxTaskCount'.`);
+    }
+
+    this.addQueueScalingPolicy(this._eventsQueue, this._eventsQueueDelay);
+    for (const subsQueue of this.subscriptionQueues) {
+      if (subsQueue.queueDelay) {
+        this.addQueueScalingPolicy(subsQueue.queue, subsQueue.queueDelay);
+      } else {
+        this.addQueueScalingPolicy(subsQueue.queue, this._eventsQueueDelay);
+      }
+    }
+    this.parentService.enableAutoScalingPolicy();
+
+    this.createLambdaFunction(service);
+  }
+
+  /**
+   * This method adds a target tracking policy based on the backlog per task custom metric
+   * to the auto scaling target configured for this service.
+   *
+   * @param queue The queue for which backlog per task metric is being configured
+   * @param queueDelay The auto scaling options for the queue
+   */
+  private addQueueScalingPolicy(queue: sqs.IQueue, queueDelay: QueueAutoScalingOptions) {
+    const acceptableBacklog = (queueDelay.acceptableLatency.toSeconds())/(queueDelay.messageProcessingTime.toSeconds());
```

Comment:
Should we be doing any validation where acceptableLantency >= messageProcessingTime?